### PR TITLE
fix: tick model generate decent granularity

### PR DESCRIPTION
**Pull request by bao-ho**
## Description
- Fix: #248 
- Root cause: default autoFormat formatter of Picasso only generates a maximum of 2 decimal digits, which is not enough when the range is very narrow.
- The fix: create a format pattern from range, allowing as many decimal digits as necessary. Assign this pattern for formatters of X-axis and Y-axis.
- Files changed:
  - [src/hooks/view-state.js](https://github.com/qlik-oss/sn-scatter-plot/pull/254/files#diff-0bbff74d9a79fd8c4daab7f9a562ef14b05866ae75e72aefbf9037bc3821c8ef): in the new code, in tick model, we get the formatters directly from chart. Therefore, there's no need to update tick-model formatters.
  - [src/models/chart-model/format-pattern-from-range.js](https://github.com/qlik-oss/sn-scatter-plot/pull/254/files#diff-f2d4aa39fecd085b9c2872e38df1356bb1eaf4d23004871a940a6477c73a0d61): create a format pattern based on the homestate data view range.
  - [src/models/chart-model/index.js](src/models/chart-model/index.js): in the new code, in tick model, we get the formatters directly from chart. Therefore, there's no need to getFormatter from chart model. Instead, we need an API function for getting format pattern based on range.
  - [src/models/tick-model/index.js](src/models/tick-model/index.js): main code change is in here. First, we can get rid of the closure object formatters, since we now get formatters directly from chat, and also there is not need for the `command.updateFormatters` API function. In the `resolve` function, we only get the format pattern once and only when autoFormat is true. However, we need to update the pattern for the chart's x and y formatters everytime we need to resolve new ticks (and display new tick labels on axes) because new formatters are constructed every time the chart is updated.
## Verification
image

### CODE REVIEW

File: src/models/chart-model/format-pattern-from-range.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  let min;
+  let max;
+  if (scaleName === KEYS.SCALE.X) {
+    ({ xAxisMin: min, xAxisMax: max } = viewHandler.getMeta().homeStateDataView);
+  } else {
+    ({ yAxisMin: min, yAxisMax: max } = viewHandler.getMeta().homeStateDataView);
+  }
+  if (!min || !max) {
+    return false;
+  }
```
Comment (T-Wizard) [inline src/models/chart-model/format-pattern-from-range.js:13]: should it return false here when min or max is 0?

File: src/models/tick-model/index.js [hunk 2]
```diff
@@ -45,16 +44,27 @@ export default function createTickModel({
...
-    x: chartModel.query.getFormatter(KEYS.FIELDS.X),
-    y: chartModel.query.getFormatter(KEYS.FIELDS.Y),
+  const formatPatterns = {
+    [KEYS.SCALE.X]: undefined,
+    [KEYS.SCALE.Y]: undefined,
   };
 
   function resolve(axis, prop) {
+    if (!formatPatterns[axis]) {
+      const field = axis === KEYS.SCALE.X ? 'qMeasureInfo.0' : 'qMeasureInfo.1';
```
Comment (T-Wizard) [inline src/models/tick-model/index.js:54]: maybe
```suggestion
      const field = axis === KEYS.SCALE.X ? KEYS.FIELDS.X : KEYS.FIELDS.Y;
```

**Comment by T-Wizard** [comment]
the code assumes that `KEYS.SCALE.X === KEYS.FORMATTER.X` (and the same for y)
(the old code did that same thing and as it is true it works form now)

**Review by T-Wizard — APPROVED** [review]

File: src/models/chart-model/format-pattern-from-range.js (hunk 1 above)
Comment (bao-ho) [reply src/models/chart-model/format-pattern-from-range.js:13]: This is a silly mistake to make.
I will to change it to ` if (min !== undefined || max !== undefined) {
    return false;
  }`. 
  Do you have a better suggestion?

File: src/models/tick-model/index.js (hunk 2 above)
Comment (bao-ho) [reply src/models/tick-model/index.js:54]: I thought about that. The problem is `KEYS.FIELDS.X = 'qMeasureInfo/0'`, while for layoutService, we need to call: `layoutService.getHyperCubeValue('qMeasureInfo.0.qIsAutoFormat, false);` (with a dot, not a slash)

**Comment by bao-ho** [comment]
> the code assumes that `KEYS.SCALE.X === KEYS.FORMATTER.X` (and the same for y) (the old code did that same thing and as it is true it works form now)

That's right. I suppose this doesn't require any fix/action?

File: src/models/chart-model/format-pattern-from-range.js (hunk 1 above)
Comment (T-Wizard) [reply src/models/chart-model/format-pattern-from-range.js:13]: can min or max be null? or is it only undefined that you need to test for.
(other values that could be relevant is NaN and Infinity)

File: src/models/chart-model/format-pattern-from-range.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  let p = '';
+  const minMagnitude = Number(Number(min).toExponential().split('e')[1]);
```
Comment (T-Wizard) [inline src/models/chart-model/format-pattern-from-range.js:15]: alternative without string manipulation
```suggestion
  const minMagnitude = Math.floor(Math.log10(Math.abs(min)));
```

File: src/models/chart-model/format-pattern-from-range.js (hunk 1 above)
Comment (bao-ho) [reply src/models/chart-model/format-pattern-from-range.js:13]: It seems that test function `Number.isFinite` will work, as it only returns `true` when min/max is an actual number.

**Pull request merged**